File: java-pi-pantry/src/main/java/com/langrsoft/pi/pantry/ExpirationDataCalculator.java
package com.langrsoft.pi.pantry;

import java.time.LocalDate;

public class ExpirationDataCalculator {
    private ShelfLifeRepository shelfLifeData;

    public ExpirationDataCalculator(ShelfLifeRepository shelfLifeData) {
        this.shelfLifeData = shelfLifeData;
    }

    public LocalDate getExpirationDate(Item item) {
        if (item.getExpirationDate() != null)
            return item.getExpirationDate();
        if (item.getCategory() == null)
            return LocalDate.MAX;
        return getExpirationDate(shelfLifeData.get(item.getCategory()), item.getSellByDate());
    }

    public LocalDate getExpirationDate(ShelfLife shelfLife, LocalDate sellByDate) {
        return sellByDate.plusDays(shelfLife.getRefrigerated());
    }
}

File: csharp-pi-pantry/csharp-pi-pantry/Domain/ExpirationDataCalculator.cs
using System;

namespace Pipantry.Domain {
    public class ExpirationDataCalculator {
        private ShelfLifeRepository shelfLifeData;

        public ExpirationDataCalculator(ShelfLifeRepository shelfLifeData) {
            this.shelfLifeData = shelfLifeData;
        }

        public DateTime ExpirationDate(Item item) {
            if (item.ExpirationDate != DateTime.MinValue)
                return item.ExpirationDate;
            if (item.Category == null)
                return DateTime.MaxValue;
            return ExpirationDate(shelfLifeData[item.Category], item.SellByDate);
        }

        private DateTime ExpirationDate(ShelfLife shelfLife, DateTime sellByDate) {
            return sellByDate.AddDays(shelfLife.Refrigerated);
        }
    }
}
